frontend/widgets/OBSBasic.cpp

```diff
@@ -1899,6 +1820,159 @@ config_t *OBSBasic::Config() const
+	config_save_safe(App()->GetUserConfig(), "tmp", nullptr);
+}
+
+bool OBSBasic::isReadyToClose()
+{
+	/* Wait for multitrack video stream to start/finish processing in the background */
+	if (setupStreamingGuard.valid() &&
+	    setupStreamingGuard.wait_for(std::chrono::seconds{0}) != std::future_status::ready) {
+		return false;
+	}
+
+	/* Do not close window if inside of a temporary event loop because we
+	 * could be inside of an Auth::LoadUI call.  Keep trying once per
+	 * second until we've exit any known sub-loops. */
+	if (os_atomic_load_long(&insideEventLoop) != 0) {
+		return false;
+	}
+
+#ifdef YOUTUBE_ENABLED
+	/* Also don't close the window if the youtube stream check is active */
+	if (youtubeStreamCheckThread) {
+		return false;
+	}
+#endif
+
+	return true;
+}
+
+bool OBSBasic::shouldPromptForClose()
+{
+	bool confirmOnExit = config_get_bool(App()->GetUserConfig(), "General", "ConfirmOnExit");
+	if (confirmOnExit && outputHandler && outputHandler->Active() && !clearingFailed) {
+		return true;
+	}
+
+	return false;
+}
+
+bool OBSBasic::promptToClose()
+{
+	isClosePromptOpen = true;
+
+	SetShowing(true);
+	QMessageBox::StandardButton button =
+		OBSMessageBox::question(this, QTStr("ConfirmExit.Title"), QTStr("ConfirmExit.Text"));
+
+	if (button == QMessageBox::No) {
+		isClosePromptOpen = false;
+		return false;
+	}
+
+	isClosePromptOpen = false;
+	return true;
+}
+
+void OBSBasic::closeWindow()
+{
+	if (isClosing()) {
+		return;
+	}
+
+	blog(LOG_INFO, SHUTDOWN_SEPARATOR);
+
+	isClosing_ = true;
+
+	/* While closing, a resize event to OBSQTDisplay could be triggered.
+	 * The graphics thread on macOS dispatches a lambda function to be
+	 * executed asynchronously in the main thread. However, the display is
+	 * sometimes deleted before the lambda function is actually executed.
+	 * To avoid such a case, destroy displays earlier than others such as
+	 * deleting browser docks. */
+	ui->preview->DestroyDisplay();
+	if (program)
+		program->DestroyDisplay();
+
+	if (outputHandler->VirtualCamActive())
+		outputHandler->StopVirtualCam();
+
+	if (introCheckThread)
+		introCheckThread->wait();
+	if (whatsNewInitThread)
+		whatsNewInitThread->wait();
+	if (updateCheckThread)
+		updateCheckThread->wait();
+	if (logUploadThread)
+		logUploadThread->wait();
+	if (devicePropertiesThread && devicePropertiesThread->isRunning()) {
+		devicePropertiesThread->wait();
+		devicePropertiesThread.reset();
+	}
+
+	QApplication::sendPostedEvents(nullptr);
+
+	signalHandlers.clear();
+	delete extraBrowsers;
+
+	saveAll();
+
+	auth.reset();
+
+#ifdef BROWSER_AVAILABLE
+	ClearExtraBrowserDocks();
+#endif
+
+	OnEvent(OBS_FRONTEND_EVENT_SCRIPTING_SHUTDOWN);
+
+	disableSaving++;
+
+	/* Clear all scene data (dialogs, widgets, widget sub-items, scenes,
+	 * sources, etc) so that all references are released before shutdown */
+	ClearSceneData();
+
+	OnEvent(OBS_FRONTEND_EVENT_EXIT);
```

Comment:
Do we need to move this earlier if plugins change OBS config or scene date? It's documented as being the last chance to save data, but we've already saved things earlier.